File: web/tests/helpers/bruteForce.ts
import type { SupportCard, TrainingPlan } from '../../src/types/models';
import { scoreDeck, type ScoreOptions } from './scoreDeck';

/**
 * 総当たり最適オラクル (合成・小プール専用)。
 * スロット制約を満たす全カード組合せを列挙し scoreDeck で採点して、真の最大を返す。
 */

const WILDCARD_TYPES = new Set(['all', 'as']);
/** 組合せ爆発ガード: これを超える候補数なら fail させる。 */
const MAX_COMBINATIONS = 20000;

function nCk(n: number, k: number): number {
  if (k < 0 || k > n) return 0;
  let r = 1;
  for (let i = 0; i < k; i++) r = (r * (n - i)) / (i + 1);
  return Math.round(r);
}

/** size 個の組合せを列挙 (インデックス集合)。 */
function* combinations(n: number, size: number): Generator<number[]> {
  const idx = Array.from({ length: size }, (_, i) => i);
  if (size === 0) {
    yield [];
    return;
  }
  if (size > n) return;
  while (true) {
    yield idx.slice();
    let i = size - 1;
    while (i >= 0 && idx[i] === i + n - size) i--;
    if (i < 0) return;
    idx[i]++;
    for (let j = i + 1; j < size; j++) idx[j] = idx[j - 1] + 1;
  }
}

/**
 * 部分集合がスロット要件を満たせるか。
 * 各属性枠は「同属性カード または all/as ワイルド」で埋める必要がある。
 * 部分集合サイズ == 総スロット数 なので、属性枠さえ埋まれば残りはフリー枠に回せる。
 * → 実現可能 ⇔ Σ max(0, required[stat] - specific[stat]) ≤ wildcard数
 */
function isFeasible(
  cards: SupportCard[],
  cardTypeSlots: Record<string, number>,
): boolean {
  const specific: Record<string, number> = {};
  let wild = 0;
  for (const c of cards) {
    if (WILDCARD_TYPES.has(c.type)) wild += 1;
    else specific[c.type] = (specific[c.type] ?? 0) + 1;
  }
  let shortfall = 0;
  for (const [stat, req] of Object.entries(cardTypeSlots)) {
    shortfall += Math.max(0, req - (specific[stat] ?? 0));
  }
  return shortfall <= wild;
}

export interface BruteForceResult {
  bestTotal: number;
  bestDeck: SupportCard[];
  evaluated: number;
}

/**
 * 総当たりで最適デッキ (scoreDeck の cap 後合計が最大) を求める。
 *
 * @param cardTypeSlots 属性枠 (例 {vo:2, da:2})
 * @param freeSlots フリー枠数
 */
export function findOptimalDeck(
  plan: TrainingPlan,
  pool: SupportCard[],
  cardTypeSlots: Record<string, number>,
  freeSlots: number,
  mainStats: string[],
  opts: ScoreOptions = {},
): BruteForceResult {
  const totalSlots =
    Object.values(cardTypeSlots).reduce((a, b) => a + b, 0) + freeSlots;

  const combos = nCk(pool.length, totalSlots);
  if (combos > MAX_COMBINATIONS) {
    throw new Error(
      `brute force too large: C(${pool.length}, ${totalSlots}) = ${combos} > ${MAX_COMBINATIONS}`,
    );
  }

  let bestTotal = -Infinity;
  let bestDeck: SupportCard[] = [];
  let evaluated = 0;

  for (const idxs of combinations(pool.length, totalSlots)) {
    const deck = idxs.map((i) => pool[i]);
    if (!isFeasible(deck, cardTypeSlots)) continue;
    evaluated += 1;
    const { cappedTotal } = scoreDeck(plan, deck, mainStats, opts);
    if (cappedTotal > bestTotal) {
      bestTotal = cappedTotal;
      bestDeck = deck;
    }
  }

  return { bestTotal, bestDeck, evaluated };
}

export { isFeasible };

File: GakumasuCalc.Tests/Helpers/BruteForce.cs
using GakumasuCalc.Models;

namespace GakumasuCalc.Tests.Helpers;

/// <summary>
/// 総当たり最適オラクル (合成・小プール専用)。TS 版 bruteForce.ts と対応。
/// スロット制約を満たす全カード組合せを列挙し ScoreDeck で採点して真の最大を返す。
/// </summary>
public static class BruteForce
{
    private static readonly HashSet<string> WildcardTypes = new() { "all", "as" };
    private const int MaxCombinations = 20000;

    public sealed class Result
    {
        public int BestTotal;
        public List<SupportCard> BestDeck = new();
        public int Evaluated;
    }

    private static long NCk(int n, int k)
    {
        if (k < 0 || k > n) return 0;
        double r = 1;
        for (int i = 0; i < k; i++) r = r * (n - i) / (i + 1);
        return (long)Math.Round(r);
    }

    private static IEnumerable<int[]> Combinations(int n, int size)
    {
        if (size == 0) { yield return Array.Empty<int>(); yield break; }
        if (size > n) yield break;
        var idx = new int[size];
        for (int i = 0; i < size; i++) idx[i] = i;
        while (true)
        {
            yield return (int[])idx.Clone();
            int i = size - 1;
            while (i >= 0 && idx[i] == i + n - size) i--;
            if (i < 0) yield break;
            idx[i]++;
            for (int j = i + 1; j < size; j++) idx[j] = idx[j - 1] + 1;
        }
    }

    /// <summary>
    /// 部分集合がスロット要件を満たせるか。
    /// 実現可能 ⇔ Σ max(0, required[stat] - specific[stat]) ≤ ワイルド(all/as)数。
    /// </summary>
    private static bool IsFeasible(List<SupportCard> cards, Dictionary<string, int> cardTypeSlots)
    {
        var specific = new Dictionary<string, int>();
        int wild = 0;
        foreach (var c in cards)
        {
            if (WildcardTypes.Contains(c.Type)) wild++;
            else specific[c.Type] = specific.GetValueOrDefault(c.Type) + 1;
        }
        int shortfall = 0;
        foreach (var kv in cardTypeSlots)
            shortfall += Math.Max(0, kv.Value - specific.GetValueOrDefault(kv.Key));
        return shortfall <= wild;
    }

    public static Result FindOptimalDeck(
        TrainingPlan plan,
        List<SupportCard> pool,
        Dictionary<string, int> cardTypeSlots,
        int freeSlots,
        List<string> mainStats,
        Dictionary<string, int>? uncapLevels = null)
    {
        int totalSlots = cardTypeSlots.Values.Sum() + freeSlots;
        long combos = NCk(pool.Count, totalSlots);
        if (combos > MaxCombinations)
            throw new InvalidOperationException(
                $"brute force too large: C({pool.Count},{totalSlots})={combos} > {MaxCombinations}");

        int bestTotal = int.MinValue;
        var bestDeck = new List<SupportCard>();
        int evaluated = 0;

        foreach (var idxs in Combinations(pool.Count, totalSlots))
        {
            var deck = idxs.Select(i => pool[i]).ToList();
            if (!IsFeasible(deck, cardTypeSlots)) continue;
            evaluated++;
            int total = DeckScorer.ScoreDeck(plan, deck, mainStats, uncapLevels);
            if (total > bestTotal)
            {
                bestTotal = total;
                bestDeck = deck;
            }
        }

        return new Result { BestTotal = bestTotal, BestDeck = bestDeck, Evaluated = evaluated };
    }
}
